pytorch_lightning/trainer/supporters.py

```diff
@@ -119,27 +122,249 @@ def mean(self):
+        """
+        if not isinstance(dataloader, DataLoader):
+            raise MisconfigurationException('Autoid only works with torch dataloaders or derived classes!')
+
+        if isinstance(dataloader.dataset, IterableDataset):
+            return dataloader
+
+        skip_keys = ['sampler', 'batch_sampler', 'dataset_kind']
+        skip_valid_keys = ['args', 'kwargs', 'self']
+
+        params = {k:v for k, v in vars(dataloader).items() if not k.startswith("_")}
+
+        valid_kwargs = set(inspect.signature(dataloader.__init__).parameters)
+        contains_dataset = True
+
+        if type(dataloader) is not DataLoader:
+            contains_dataset = "dataset" in valid_kwargs
+            valid_kwargs.update(inspect.signature(DataLoader.__init__).parameters)
+
+        dl_args = {
+            name: params[name] for name in valid_kwargs
+            if name in params and name not in skip_keys
+        }
+
+        multiprocessing_context = dataloader.multiprocessing_context
+
+        # override parameters to enable batch_sampler injection
+        dl_args["batch_size"] = 1
+        dl_args["sampler"] = None
+        dl_args["shuffle"] = None
+        dl_args["batch_sampler"] = LightningBatchSamplerWrapper(dataloader.batch_sampler)
+        dl_args["drop_last"] = False
+        dl_args['multiprocessing_context'] = multiprocessing_context
+
+        missing_kwargs = valid_kwargs.difference(skip_valid_keys).difference(set(dl_args))
+        if len(missing_kwargs) != 0:
+            dataloader_cls_name = dataloader.__class__.__name__
+            rank_zero_warn(
+                f"Trying to replace to wrap your BatchSampler in your {dataloader_cls_name} dataloader."
+                "This would fail as your DataLoader doesn't expose as attributes all its __init__ parameters. "
+                f"Missing attributes are {missing_kwargs}"
+                "HINT: use Trainer(enable_predict_auto_id=False) and provide your own id. "
+                "Check out the doc for Testing. ", UserWarning
+            )
+            return dataloader
+
+        if not contains_dataset:
+            dl_args.pop('dataset')
+
+        dataloader = type(dataloader)(**dl_args)
+        dataloader.multiprocessing_context = multiprocessing_context
+        return dataloader
+
+
 class PredictionCollection(object):
-    def __init__(self, global_rank: int, world_size: int):
+
+    """
+    This class is used to collect predictions.
+
+    The legacy API built using the following functions:
+        - add
+        - _add_prediction
+        - to_disk
+
+    The new API built using the following functions:
+        - cache: Receive a new predictions to cache. Handle validation
+        - _cache_prediction: Handle storing
+        - attach_predictions: When the epoch is finished, add predictions to results object
+        - all_gather_predictions: Gather predictions accross multiple processses.
```

Comment:
I feel confused about what is the difference (it is not described in docs) between `cache` and `attach_predictions` as both are public API when each shall be used...
so to say the `cache` add some particular predictions to some stach related to batch and when all prediction are collected, at the end of test some post-processing (`attach_predictions`) is called?
- can we describe the flow in docs?
- the word `cache` is a bit confusing I would suggest `append` as more intuitive
- I would rename `attach_predictions` as `finalize_predictions`

Reply:
Done !